packages/react-graphql-universal-provider/src/GraphQLUniversalProvider.tsx

```diff
@@ -1,33 +1,64 @@
 import React from 'react';
 
+import ApolloClient, {ApolloClientOptions} from 'apollo-client';
 import {useSerialized} from '@shopify/react-html';
-import {ApolloProvider, createSsrExtractableLink} from '@shopify/react-graphql';
+import {
+  ApolloProvider,
+  createSsrExtractableLink,
+  createOperationDetailsLink,
+  GraphQLOperationDetails,
+} from '@shopify/react-graphql';
 import {useLazyRef} from '@shopify/react-hooks';
 import {createApolloBridge} from '@shopify/react-effect-apollo';
+import {useNetworkManager} from '@shopify/react-network';
+
+export type GraphQLClientOptions = ApolloClientOptions<any>;
+export const GRAPHQL_OPERATIONS = Symbol('graphQLOperations');
 
 interface Props {
   children?: React.ReactNode;
-  createClient(): import('apollo-client').ApolloClient<any>;
+  createClientOptions(): ApolloClientOptions<any>;
 }
 
-export function GraphQLUniversalProvider({children, createClient}: Props) {
+export function GraphQLUniversalProvider({
+  children,
+  createClientOptions,
+}: Props) {
   const [initialData, Serialize] = useSerialized<object | undefined>('apollo');
+  const network = useNetworkManager();
 
   const [client, link] = useLazyRef<
     [
       import('apollo-client').ApolloClient<any>,
       ReturnType<typeof createSsrExtractableLink>
     ]
   >(() => {
+    const clientOptions = createClientOptions();
     const link = createSsrExtractableLink();
-    const client = createClient();
-    client.link = link.concat(client.link);
+
+    if (clientOptions.link) {
+      clientOptions.link = link
+        .concat(
+          createOperationDetailsLink({
+            onOperation(operation) {
+              if (network) {
+                const newServerState = addOperationToServerState(
```

Comment:
after which I wonder if we can use `useNetworkEffect` instead of `useNetworkManager`?
and if that is possible, maybe we can have `useGraphqlOperation` in `react-network` to encapsulate the network related things.

Reply:
>  maybe we can have useGraphqlOperation in

In the end, we need to be able to use these graphql operations in a koa middleware. We can't call `useGraphqlOperation` in the koa middleware. Unless there's a pattern here that I'm not thinking of.

Reply:
no, hook have to be use in a react component. Can you remind me why this is have to be use in koa middleware and not in the App code?

Reply:
> why this is have to be use in koa middleware

Because we need to send off the [metrics](https://github.com/Shopify/quilt/pull/1018/files#diff-167e62803c79a1e1b024bbead47ca47bR36-R40) in  the [response header](https://github.com/Shopify/quilt/pull/1018/files#diff-167e62803c79a1e1b024bbead47ca47bR42-R43) to the rails server.